racket/collects/racket/match/runtime.rkt

```diff
@@ -69,3 +76,21 @@
                 (syntax-column stx)
                 (syntax-position stx)
                 (syntax-span stx))))
+
+(define undef (gensym))
+(define user-def (gensym))
+
+(define (undef? v)
+  (eq? undef v))
+
+(define (user-def? v)
```

Comment:
I don't understand the role of `user-def` in this.

Reply:
When you use `(hash* [1 a #:default 3])`, if the hash table value doesn't have key 1, `a` is matched against `3`.

Now, consider `(hash* [1 a #:default (begin (println 'foo) 3)] [2 3])`. What should happen regarding `(println 'foo)` in the following cases?

- The hash table has key `1`
- The hash table does not have have key `1` and `2`.

The strategy that I currently use is to evaluate `(println 'foo)` only when the pattern is really matched and the corresponding key really doesn't exist. `user-def` is a dummy value to facilitate this delayed evaluation.

Of course, other strategies are possible too. I.e., always evaluate `(println 'foo)` whether or not key exists, and whether or not the clause is matched. That would be simpler to implement, but I personally don't like that, primarily because the default value might be expensive to compute. But you might have different opinion.

More generally, this is related to the order of evaluation among expressions in the pattern. Right now I try to be as lazy as possible, where keys are first evaluated, and then default values. But another possibility is to interleave key and default value evaluations.

Reply:
> *The strategy that I currently use is to evaluate `(println 'foo)` only when the pattern is really matched and the corresponding key really doesn't exist.*

That's strange. What do other parts of `match` do with expressions with side effects? 

Another stranger case:

    (hash* [1 a #:default (begin (println 'foo) (random 7))] [2 3])

Reply:
@gus-massa this issue is trivial in other patterns, because there are at most one expression in other patterns.

`match` has many short-circuiting opportunities, so generally expressions are only evaluated when needed. E.g.

```
(define (f)
  (println "foobar")
  (λ (v) #t))

(match 1
  [(and 2 (app (f))) 'matched]
  [_ 'not-matched])
```

doesn't print `"foobar"`.

---------------------

Your example is a good one. Should `random` be called once or multiple times per site? This is relevant when there's a `match` backtracking that tries the same pattern multiple times. But this already occurs in the current `match`. E.g.

```
#lang racket

(define (f)
  (println "foobar")
  (random 2))

(match (list (list 0) (list 1))
  [(list-no-order
    (list (== (f)))
    (list (== (f)))) 'matched]
  [_ 'not-matched])
```

can print `"foobar"` more than twice. The `hash` pattern has the same behavior.

I'm inclined to say that there's no guarantee regarding evaluation order when backtracking occurs. People rarely use it. And when they do, I think it's reasonable that they must ensure the program doesn't have anything "weird."